com.unity.multiplayer.mlapi/Runtime/Timing/NetworkTime.cs

```diff
@@ -0,0 +1,140 @@
+        /// Gets the amount of network ticks which have passed until reaching the current time value.
+        /// </summary>
+        public int Tick => m_CachedTick;
+
+        /// <summary>
+        /// Gets the tickrate of the system of this <see cref="NetworkTime"/>.
+        /// </summary>
+        public int TickRate => m_TickRate;
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="NetworkTime"/> struct.
+        /// </summary>
+        /// <param name="tickRate">The tickrate.</param>
+        public NetworkTime(int tickRate)
+        {
+            m_TickRate = tickRate;
+            m_TickInterval = 1f / m_TickRate; // potential floating point precision issue, could result in different interval on different machines
+            m_CachedTickOffset = 0;
+            m_CachedTick = 0;
+            m_Time = 0;
+        }
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="NetworkTime"/> struct.
+        /// </summary>
+        /// <param name="tickRate">The tickrate.</param>
+        /// <param name="tick">The time will be created with a value where this many tick have already passed.</param>
+        /// <param name="tickOffset">Can be used to create a <see cref="NetworkTime"/> with a non fixed time value by adding an offset to the given tick value.</param>
+        public NetworkTime(int tickRate, int tick, double tickOffset = 0d)
+            : this(tickRate)
+        {
+            Assert.IsTrue(tickOffset < 1d / tickRate);
+            this += tick * m_TickInterval + tickOffset;
+        }
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="NetworkTime"/> struct.
+        /// </summary>
+        /// <param name="tickRate">The tickrate.</param>
+        /// <param name="time">The time value as a float.</param>
+        public NetworkTime(int tickRate, double time)
+            : this(tickRate)
+        {
+            this += time;
+        }
+
+
+        /// <summary>
+        /// Converts the network time into a fixed time value.
+        /// </summary>
+        /// <returns>A <see cref="NetworkTime"/> where Time is the FixedTime value of this instance.</returns>
+        public NetworkTime ToFixedTime()
+        {
+            return new NetworkTime(m_TickRate, m_CachedTick);
+        }
+
+        private void UpdateCache()
+        {
+            double d = m_Time / m_TickInterval;
+            m_CachedTick = (int)d;
+            m_CachedTickOffset = ((d - Math.Truncate(d)) * m_TickInterval);
+
+            // This handles negative time, decreases tick by 1 and makes offset positive.
+            if (m_CachedTick < 0 && m_CachedTickOffset != 0d)
+            {
+                m_CachedTick--;
+                m_CachedTickOffset = m_TickInterval + m_CachedTickOffset;
+            }
+        }
+
+        public static NetworkTime operator -(NetworkTime a, NetworkTime b)
+        {
+            return new NetworkTime(a.TickRate, a.Time - b.Time);
```

Comment:
I'd be curious to see if the compiler actually does an allocation with struct operator overload here or if one of the operands' storage is used.  If the former I would probably want to look at how many of these get created in a tick

Reply:
It will stackalloc a new struct. There is no way to make a struct return not create a stackalloc. The same happens if you do `Vector3 a + Vector3b`, I don't see why this is an issue.